JavaScript/StaticTypeChecking/README.md

```diff
@@ -0,0 +1,128 @@
+        );
+    }
+}
+```
+ We bind the method `onButtonSelect` to `this` in the constructor. The benefit of binding in the constructor vs in the render method is we don't rebind `this` on every click.
+
+ Let's add flow to the file.
+
+ ```Javascript
+// @flow
+type Props = {
+    text: string
+}
+
+class BindingExample extends React.Component {
+    state: {mySafeSpace: boolean}
+
+    constructor(props: Props) {
+        super(props);
+        this.state = {
+            mySafeSpace: true
+        }
+
+        (this:any).onTriggered = this.onTriggered.bind(this);
+    }
+
+    onTriggered() {
+        this.setState{(mySafeSpace: false)};
+    }
+
+    render() {
+        return (
+            <button onClick={this.onTriggered}>
+                {this.props.text}
+            </button>
+        );
+    }
+}
+```
+
+Sweet. Not too bad. but wait...
+Why the need for `(this: any)` in the constructor now? I thought we were supposed to limit the instances of `any`? Without saying `this` can be anything, flow will throw an error referencing 'contravariance' and 'covariance' as the root of the type error. [Here](http://www.i-programmer.info/programming/theory/1632-covariance-and-contravariance-a-simple-guide.html) [are](http://tomasp.net/blog/variance-explained.aspx/) [some](https://msdn.microsoft.com/en-us/library/dd799517(v=vs.110).aspx) resources on contravariance and covariance in category theory. It's a really brilliant idea; the reason why flow typing is superior than the static typing of that in other languages like Java. But whether or not each dev should understand the theory necessary in order to fix the error in the codebase and understand _why_ is debatable.
+
+Once you start introducing paradigms like Higher Order Components (in React) you'll start running into more complex types and will really need to understand more about type theory. Not a _bad_ thing, since we're devs and our whole career is learning. However, weighing the pros and cons of implementing this paradigm is something that should be done on a project to project basis based on the skill level of the team. If you have an understanding of type theory or want to learn it, and feel you can lead a team through these errors, then go for it. The real gotcha is that it has the possibility of both speeding up _or_ hindering your development.
+
+## Conclusion
+
+Flow is a new kid on the block. There are new pull requests being opened constantly which are fixing ailments like the one just mentioned. An application - even when perfectly and specifically typed - can still fail when used with third party libraries. There are typings for popular libraries like React, but there is not a type safety guarantee when your application calls upon methods and apis from libraries without set types. You can create those yourself, but the tradeoff is that it takes time. However, there are immense benefits to making the application structure better defined with more contractual agreements.
```

Comment:
`Class` is an example. http://www.saltycrane.com/blog/2016/06/flow-type-cheat-sheet/